pkg/codegen/hcl2/model/type_object.go

```diff
@@ -89,6 +117,15 @@ func (t *ObjectType) equals(other Type, seen map[Type]struct{}) bool {
 	if t == other {
 		return true
 	}
+
+	otherObject, ok := other.(*ObjectType)
+	if !ok {
+		return false
+	}
+	if len(t.Properties) != len(otherObject.Properties) {
+		return false
+	}
+
```

Comment:
There is something slightly spooky semantically here (before the changes). The function trying to prove `t.equals(x)` decides that it's true if it needs to prove `t.equals(y)` goal as part of the proof...

It sounds to me it would prove `A=B` from:

```
type A struct {
    x: int
    y: A
}

type B struct {
    x: int
    y: string
}
```

Equality on these infinite coinductive structures is tricky. I wonder if I am missing something and this actually works, or we just do not hit corner cases in practice.

Reply:
I do think that this is subtly broken, and has been since we started handling recursive types. What we really want to say is that the following two types are equal:

```
struct {
    x int
    y self
}

struct {
    x int
    y self
}
```

Note that I'm deliberately avoiding the use of names here, as the type system doesn't give names to object types. I considered making that change, but it's quite a bit more disruptive. It might be worth doing, however, as it may reduce the equality problem to "just check the names".